src/morale.cpp

```diff
@@ -137,10 +218,379 @@ void morale_point::proceed( int ticks )
+    apply_temperature( p );
+    // Invalidate level to recalculate it on demand
+    invalidate();
+}
+
+void player_morale::display( const player *p )
+{
+    // Create and draw the window itself.
+    WINDOW *w = newwin( FULL_SCREEN_HEIGHT, FULL_SCREEN_WIDTH,
+                        ( TERMY > FULL_SCREEN_HEIGHT ) ? ( TERMY - FULL_SCREEN_HEIGHT ) / 2 : 0,
+                        ( TERMX > FULL_SCREEN_WIDTH ) ? ( TERMX - FULL_SCREEN_WIDTH ) / 2 : 0 );
+    draw_border( w );
+
+    // Figure out how wide the name column needs to be.
+    int name_column_width = 18;
+    for( auto &i : points ) {
+        int length = utf8_width( i.get_name() );
+        if( length > name_column_width ) {
+            name_column_width = length;
+        }
+    }
 
-    if( is_expired() ) {
-        bonus = 0;
-    } else if( age > decay_start ) {
-        bonus *= logarithmic_range( decay_start, duration, age );
+    // If it's too wide, truncate.
+    if( name_column_width > 72 ) {
+        name_column_width = 72;
+    }
+
+    // Start printing the number right after the name column.
+    // We'll right-justify it later.
+    int number_pos = name_column_width + 1;
+
+    // Header
+    mvwprintz( w, 1,  1, c_white, _( "Morale Modifiers:" ) );
+    mvwprintz( w, 2,  1, c_ltgray, _( "Name" ) );
+    mvwprintz( w, 2, name_column_width + 2, c_ltgray, _( "Value" ) );
+
+    // Ensure the player's persistent morale effects are up-to-date.
+    apply_persistent( p );
+
+    const morale_mult mult = get_traits_mult( p );
+    // Print out the morale entries.
+    for( size_t i = 0; i < points.size(); i++ ) {
+        std::string name = points[i].get_name();
+        int bonus = points[i].get_net_bonus( mult );
+
+        // Print out the name.
+        trim_and_print( w, i + 3,  1, name_column_width, ( bonus < 0 ? c_red : c_green ), name.c_str() );
+
+        // Print out the number, right-justified.
+        mvwprintz( w, i + 3, number_pos, ( bonus < 0 ? c_red : c_green ),
+                   "% 6d", bonus );
+    }
+
+    // Print out the total morale, right-justified.
+    int mor = get_level( p );
+    mvwprintz( w, 20, 1, ( mor < 0 ? c_red : c_green ), _( "Total:" ) );
+    mvwprintz( w, 20, number_pos, ( mor < 0 ? c_red : c_green ), "% 6d", mor );
+
+    // Print out the focus gain rate, right-justified.
+    double gain = ( p->calc_focus_equilibrium() - p->focus_pool ) / 100.0;
+    mvwprintz( w, 22, 1, ( gain < 0 ? c_red : c_green ), _( "Focus gain:" ) );
+    mvwprintz( w, 22, number_pos - 3, ( gain < 0 ? c_red : c_green ), _( "%6.2f per minute" ), gain );
+
+    // Make sure the changes are shown.
+    wrefresh( w );
+
+    // Wait for any keystroke.
+    getch();
+
+    // Close the window.
+    werase( w );
+    delwin( w );
+}
+
+void player_morale::clear()
+{
+    points.clear();
+    invalidate();
+}
+
+void player_morale::invalidate()
+{
+    level_is_valid = false;
+}
+
+int player_morale::get_hoarder_penalty( const player *p ) const
```

Comment:
You should use references (const if possible) instead of pointers. Your code seems to assume the pointer is always non-null (no checks for null). Using references in that situation makes that clear to the caller of those functions.